File: C++/BOJ/4949.cpp
#include<stack>
#include<iostream>
#include<stdio.h>
#include<string>
#include<map>
using namespace std;
int main()
{
    map<char, char> dic;
    dic[')'] = '(';
    dic[']'] = '[';
    while (1)
    {
        string str;
        stack<char> l_stack;
        stack<char> r_stack;
        cin>>str;
        //scanf("%s",str);
        if (str == ".")
            break;
        for (auto c : str)
        {
            if (c == '[' || c == '(')
                l_stack.push(c);
            if (c == ']' || c == ')')
            {
                r_stack.push(c);
                if (l_stack.top() == dic[r_stack.top()])
                {
                    l_stack.pop();
                    r_stack.pop();
                }
            }
        }
        if (l_stack.size() == 0 && r_stack.size() == 0)
            cout<<"yes\n";
        else
            cout<<"no\n";
    }
}

File: Python/BOJ/4949.py
import sys

def solution():
    dic={')': '(', ']': '['}
    while (1):
        l_stack = []
        r_stack = []
        top = 'e'
        string = input()
        if string == ".":
            break
        for ch in string:
            if ch in ['(', '[']:
                l_stack.append(ch)
                top = ch
            if ch in dic:
                r_stack.append(ch)
                if top == dic[ch]:
                    l_stack.pop()
                    r_stack.pop()
                    if len(l_stack) > 0:
                        top = l_stack[-1]
                    else:
                        top = 'e'
        if len(l_stack) == 0 and len(r_stack) == 0:
            print("yes")
        else:
            print("no")

solution()